zapcore/write_syncer_test.go

```diff
@@ -18,110 +18,95 @@
-	ws := MultiWriteSyncer(AddSync(failer), AddSync(second))
-
-	_, err := ws.Write([]byte("fail"))
-	assert.Error(t, err, "Expected error from call to a writer that failed")
-	assert.Equal(t, []byte("fail"), second.Bytes(), "Expected second sink to be written after first error")
-}
-
-func TestMultiWriteSyncerSync_PropagatesErrors(t *testing.T) {
-	badsink := &testutils.Buffer{}
-	badsink.SetError(errors.New("sink is full"))
-	ws := MultiWriteSyncer(&testutils.Discarder{}, badsink)
-
-	assert.Error(t, ws.Sync(), "Expected sync error to propagate")
-}
-
-func TestMultiWriteSyncerSync_NoErrorsOnDiscard(t *testing.T) {
-	ws := MultiWriteSyncer(&testutils.Discarder{})
-	assert.NoError(t, ws.Sync(), "Expected error-free sync to /dev/null")
-}
-
-func TestMultiWriteSyncerSync_AllCalled(t *testing.T) {
-	failed, second := &testutils.Buffer{}, &testutils.Buffer{}
-
-	failed.SetError(errors.New("disposal broken"))
-	ws := MultiWriteSyncer(failed, second)
-
-	assert.Error(t, ws.Sync(), "Expected first sink to fail")
-	assert.True(t, failed.Called(), "Expected first sink to have Sync method called.")
-	assert.True(t, second.Called(), "Expected call to Sync even with first failure.")
+func TestBufferedWriterPropagatesFailures(t *testing.T) {
+	failSyncer := &testutils.Buffer{}
+	failSyncer.SetError(errors.New("no sync"))
+
+	failWriteSyncer := &testutils.FailWriter{}
+	failWriteSyncer.SetError(errors.New("no sync"))
+
+	tests := []struct {
+		desc           string
+		initial        string
+		try            string
+		ws             WriteSyncer
+		n              int
+		expectWriteErr string
+		expectSyncErr  string
+	}{
+		{
+			desc: "no write, no buffered content, fail writes",
+			ws:   &testutils.FailWriter{},
+		},
+		{
+			desc:          "no write, some buffered content, fail writes",
+			initial:       "ab",
+			ws:            &testutils.FailWriter{},
+			expectSyncErr: "failed",
+		},
+		{
+			desc:           "write overflows buffer, fail writes",
+			initial:        "ab",
+			try:            "cde",
+			ws:             &testutils.FailWriter{},
+			n:              3,
+			expectWriteErr: "failed",
+			expectSyncErr:  "failed",
+		},
+		{
+			desc:           "write can't be buffered, fail writes",
+			initial:        "ab",
+			try:            "cdefg",
+			ws:             &testutils.FailWriter{},
+			n:              5,
+			expectWriteErr: "failed; failed",
+		},
+		{
+			desc:          "no write, no buffered content, fail syncs",
+			ws:            failSyncer,
+			expectSyncErr: "no sync",
+		},
+		{
+			desc:           "write overflows buffer, fail writes and syncs",
+			initial:        "ab",
+			try:            "cde",
+			ws:             failWriteSyncer,
+			n:              3,
+			expectWriteErr: "failed",
+			expectSyncErr:  "failed; no sync",
+		},
+	}
+
+	for _, tt := range tests {
+		buf := Buffer(4, tt.ws)
+		_, err := buf.Write([]byte(tt.initial))
+		assert.NoError(t, err, "Unexpected error writing initial buffer contents (%q).", tt.desc)
```

Comment:
Probably worth using subtests (my heuristic is "at the point you feel compelled to interpolate a desc string") ;-)